### ISSUE TITLE
No reconnection if "Transport endpoint not connected"

### ISSUE THREAD

**Issue by mjramtech** [libfuse/sshfs#202]
I suspect this issue is known/understood, but I can't find an open issue for it.

This has been a problem for me for some time now, and chasing threads on Stack Overflow and the like (search sshfs "Transport endpoint is not connected") suggests to me this is a years old issue that has never been addressed. Whether the fix would be in sshfs or elsewhere I don't know, but I think it's fairly reasonable to expect a reconnection attempt to be made if the reconnect option has been passed and the mount point is no longer available.

The problem appears to be when the mount is still listed under `mount`, but there is no process for it under sshfs, e.g.:

Note that while I've changed some users/paths for anonymity, the issue is otherwise exactly as described, i.e. both mount points, the one currently working and the one currently not, are adjacent on both server and client (so you might expect them to either work or fail together, but they don't).

Autofs config:

`cat /etc/auto.sshfs`
`data1 -fstype=fuse,port=22,rw,nodev,nonempty,noatime,allow_other,max_read=65536,reconnect,workaround=all,ServerAliveInterval=15,ServerAliveCountMax=3 :sshfs\#user@server1\:/storage/data1`
`data2 -fstype=fuse,port=22,rw,nodev,nonempty,noatime,allow_other,max_read=65536,reconnect,workaround=all,ServerAliveInterval=15,ServerAliveCountMax=3 :sshfs\#user@server1\:/storage/data2`

Output of mount command:

`mount | grep server1`
`user@server1:/storage/data1 on /mnt/sshfs/data1 type fuse.sshfs (rw,nodev,noatime,user_id=0,group_id=0,allow_other)`
`user@server1:/storage/data2 on /mnt/sshfs/data2 type fuse.sshfs (rw,nodev,noatime,user_id=0,group_id=0,allow_other)`

Autofs status (note absence of data2 mount point specified above):

user@serverx:/home/user# systemctl status autofs
● autofs.service - Automounts filesystems on demand
   Loaded: loaded (/lib/systemd/system/autofs.service; enabled; vendor preset: enabled)
   Active: active (running) since Wed 2019-10-23 14:46:52 UTC; 2 months 13 days ago
 Main PID: 8524 (automount)
    Tasks: 17 (limit: 4915)
   Memory: 8.8M
      CPU: 18h 19min 6.124s
   CGroup: /system.slice/autofs.service
           ├─8524 /usr/sbin/automount --pid-file /var/run/autofs.pid
           ├─8718 ssh -X -a -oClearAllForwardings=yes -oport=22 -oServerAliveInterval=15 -oServerAliveCountMax=3 -2 user@server1 -s sftp
           └─8755 sshfs user@server1:/storage/data1 /mnt/sshfs/data1 -o rw,nodev,noatime,uid=0,gid=0,port=22,nonempty,allow_other,max_read=65536,reconnect,workaround=[truncated by terminal]

Now, if I `umount /mnt/sshfs/data2`, then attempt an `ls` of the same directory, the volume is automatically remounted. But without the `umount`, `ls` (here of the parent directory) yields:

user@serverx:/home/user# ls -l /mnt/sshfs/
ls: cannot access '/mnt/sshfs/data2': Transport endpoint is not connected
total 4
drwxr-xr-x 1 root root 4096 Jul 22 11:46 data1
d????????? ? ?    ?       ?            ? data2

So, why is the volume not remounted automatically, when attempts to access it are returning "not connected" errors?

**Comment by mjramtech** [comment]
I've just done a bit of experimenting, killing various elements on client/server, and the only way I can reproduce the problem I'm occasionally seeing is by killing sshfs on the client:

Kill ssh session on server: mount recovers
Kill sftp-server on server: mount recovers
kill ssh session on client: mount recovers
kill sshfs on client: have to manually `umount`, after which it again mounts on demand (as problem originally described)

So, without concrete evidence I'll suggest sshfs itself is occasionally exiting. Should autofs handle this?

**Comment by andymwood** [comment]
I'm encountering the same problem, though I'm not using autofs to mount the directories.

Server 'huey' is sshfs-mounting a couple of directories that reside on 'dewey':
```bash
[root@huey ~]# mount|grep dewey
dewey:/srv/archive on /srv/archive type fuse.sshfs (ro,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other)
dewey:/srv/records on /srv/records type fuse.sshfs (ro,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other)
```

The connection for one of the directories is no longer working:
```bash
[root@huey ~]# df | grep dewey
df: ‘/srv/archive’: Transport endpoint is not connected
dewey:/srv/records                             314399976   218728920   95671056  70% /srv/records
```

And sshfs appears to be mysteriously exiting:
```bash
[root@huey ~]# pgrep -af dewey
7263 ssh -x -a -oClearAllForwardings=yes -2 dewey -s sftp
7264 sshfs dewey:/srv/records /srv/records -o reconnect -o allow_other -o ro
# There should be a process for 'sshfs dewey:/srv/archive ...'
```

**Comment by Nikratio** [comment]
Thanks for reporting this! I believe this is the same as issue #3.

**Comment by mjramtech** [comment]
Is this definitely a duplicate of issue #3? I've seen issue #3 many times, including twice recently on a production machine, but from my perspective this is a different issue - it doesn't hang the system, it just doesn't attempt a reconnect despite the remote machine being available.

I appreciate this could be a different manifestation of the same underlying issue, it's just without any explanation to that effect I worry this could be swept under the rug of a now almost 5 year old ticket.

**Comment by Nikratio** [comment]
Apologies, I wasn't paying enough attention here. `Transport Endpoint not connected` means that SSHFS has crashed and is no longer running, so naturally it cannot reconnect. 

What's needed here would be a stacktrace and ideally core dump of SSHFS when it crashes. If you can provide this, please do re-open this ticket.

**Comment by mjramtech** [comment]
Thanks for the quick reply. Don't know that I have the necessary permissions to re-open the ticket. Hopefully myself or somebody else can capture this and post the results here. Is closed is the correct status in the meantime?

**Comment by Nikratio** [comment]
Actually, please just open a new ticket ("SSHFS crashes"). That way, we don't have the confusion about it not reconnecting :-).

**Comment by mjramtech** [comment]
If this project is as undermaintained as it seems, I sense that would be no better a use of my time than it was yours closing the ticket unfixed.

**Comment by andymwood** [comment]
In case this is useful, I had been using version 2.5 of sshfs. The [ChangeLog](https://github.com/libfuse/sshfs/blob/sshfs_2.x/ChangeLog.rst) has this:

> Release 2.10 (2017-08-03)
>
>     Fixed a crash due to a race condition when listing directory contents.

I also discovered that my scripts were using `umount` to unmount the remote directory, whereas the documentation says to use `fusermount -u`.

I upgraded sshfs and fixed my scripts a few months ago and the problem hasn't recurred since.
